src/btree/bt_walk.c

```diff
@@ -666,7 +647,56 @@ int
 __wt_tree_walk_count(WT_SESSION_IMPL *session,
     WT_REF **refp, uint64_t *walkcntp, uint32_t flags)
 {
-	return (__tree_walk_internal(session, refp, walkcntp, NULL, flags));
+	return (__tree_walk_internal(
+	    session, refp, walkcntp, NULL, NULL, flags));
+}
+
+/*
+ * __wt_tree_walk_custom_skip --
+ *	Walk the tree calling a custom function to decide whether to skip refs.
+ */
+int
+__wt_tree_walk_custom_skip(
+    WT_SESSION_IMPL *session, WT_REF **refp,
+   int (*skip_func)(WT_SESSION_IMPL *, WT_REF *, void *, bool *),
+   void *func_cookie, uint32_t flags)
+{
+	return (__tree_walk_internal(session, refp,
+	    NULL, skip_func, func_cookie, flags));
+}
+
+/*
+ * __tree_walk_skip_count_callback --
+ *	Optionally skip leaf pages.
+ * When the skip-leaf-count variable is non-zero, skip some count of leaf
+ * pages, then take the next leaf page we can.
+ *
+ * The reason to do some of this work here, is because we can look at the cell
+ * and know it's a leaf page without reading it into memory. If this page is
+ * disk-based, crack the cell to figure out it's a leaf page without reading
+ * it.
+ */
+static int
+__tree_walk_skip_count_callback(
+    WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool *skipp)
+{
+	uint64_t *skipleafcntp;
+
+	skipleafcntp = (uint64_t *)context;
+	WT_ASSERT(session, skipleafcntp != NULL);
+
+	/*
+	 * Skip deleted pages visible to us.
+	 */
+	if (ref->state == WT_REF_DELETED &&
+	    __wt_delete_page_skip(session, ref, false))
+		*skipp = true;
+	else if (*skipleafcntp > 0 && __ref_is_leaf(ref)) {
+		--*skipleafcntp;
+		*skipp = true;
+	} else
+		*skipp = false;
+	return (0);
 }
```

Comment:
I like splitting this out into its own callback function -- it's cleaner than having it interleaved with ordinary walk.